sourcegraph/sourcegraph
# bext: add `nameTransformations` support to GitLab

Closes https://github.com/sourcegraph/sourcegraph/issues/35470

If the GitLab config has [name transformation rules](https://docs.sourcegraph.com/admin/external_service/gitlab#nameTransformations) defined, matching GitLab repositories are available on the Sourcegraph instance under the name/URL modified accordingly. 
Previously Sourcegraph browser extension parsed DOM on GitLab to get the current repo name. If the name transformation rules were applied, repo names on the code host no longer matched the one on the Sourcegraph instance, causing all requests having repo name as a parameter to resolve incorrectly (usually resulting in a repo not found error).
This PR fixes this issue by getting the matching repo name from GraphQL API before adding the code intelligence to the code host. The repo name received from GraphQL API is used in requests to Sourcegraph GraphQL API, and the repo name from the code host (extracted from DOM) is still used for requests to GitLab API.

[Loom](https://www.loom.com/share/e7d10f7521364b3a8dc7d9eb54cf04fd) (_note_: GraphQL query name was changed from `RepositoryName` to `ResolveRepoName` to be aligned with other queries names)

## Test plan
1. `sg start` + `sg run bext`
2. Log in as an admin user to your instance and add a GitLab code host configuration. You may need to comment out [these lines](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4730c0988d0018924738dd71dfe77bbe5b7b29c8/-/blob/cmd/frontend/graphqlbackend/external_services.go?L54-56) to be able to add another code host config when running in dev mode.
3. Ensure that the added GitLab repositories are synced.
4. Set your instance as a Sourcegraph instance in the browser extension options menu.
5. Open a synced repository on GitLab and ensure that the code intel works.
6. Add name transformation rules to your GitLab code host config. Ensure these rules were applied: matching repos should be available on your Sourcegraph instance under the modified name. If you cannot modify the existing code host config (because of [this restriction](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4730c0988d0018924738dd71dfe77bbe5b7b29c8/-/blob/cmd/frontend/graphqlbackend/external_services.go?L139-141)), just add a new one with the name transformations defined.
7. Open repo on GitLab and ensure that the code intel features still work.


## App preview:

- [Web](https://sg-web-taras-yemets-bext-support-name.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-rzbrgvysnq.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
